Update

cbk914 opened the pull request.

a-dma committed:
Indent

a-dma committed:
Use calloc to allocate buf

a-dma committed:
Check buf length in parse_ssh_format more thoroughly

a-dma committed:
Check strdup return value when allocation attributes

a-dma committed:
Use a define for user presence

a-dma committed:
Indent

a-dma committed:
Drop unnecessary sizeof(char)

a-dma committed:
Make sure to save one \0 byte in buf

a-dma committed:
Initialize decoded_initial to NULL

a-dma committed:
Add default authfile for sshformat

a-dma committed:
Indent

a-dma committed:
Fix padding check for SSH format

a-dma committed:
Fixup old format compatibility

a-dma committed:
Add a test for parsing credential files

a-dma committed:
Indent tests

a-dma committed:
Change tests harness to run serial tests

This helps when printing results in the CI

a-dma committed:
Change ownership of the test credential files

a-dma committed:
Indent

a-dma committed:
Fix print formatting on decoded application field

a-dma committed:
Update SSH flags to latest version and attributes management

a-dma committed:
Use a fido_opt_t in pamu2fcfg

a-dma committed:
Use fido_opt_t in do_manual_authentication

a-dma committed:
Don't set uv when setting PIN

a-dma committed:
Check against "+attribute" rather "attribute"

a-dma committed:
Indent

Commit:
actions: CodeQL scanning

a-dma committed:
Add a script to generate test cases

a-dma committed:
Add a whole bunch of test cases

Parsing combination of single and multiple devices in one file with
different sets of attributes

a-dma committed:
Rework the logic to parse authentication files

Make sure that each credential is always parsed correctly including
attributes and lack thereof (for the old format).

a-dma committed:
Appease clang-format

a-dma committed:
Fix permissions on tests

This change adds a new custom suffix to automake and treats
credentials as templates. Upon running the tests actual credential
files (.cred) will be generated with the current user name.

Similarly the tests will read the user name from the environment and
use that value.

a-dma committed:
Update Travis build to not do chown

a-dma committed:
Indent

a-dma committed:
Update documentation

Reflect the changes for the transition from libu2f to libfido2.
Add some text on the SSH format.
Update manpages.

Relates to #140

a-dma committed:
Fix code block formatting

a-dma committed:
Fix dependency name

a-dma committed:
Mention minimum libfido2 version required

a-dma committed:
Replace Debian with Ubuntu in the documentation

a-dma committed:
Update version number to reflect changes

a-dma committed:
Update NEWS file

a-dma committed:
Fix typo

a-dma committed:
User autoconf instead of make to generate test input files

a-dma committed:
Use correct filename in Makefile.am

a-dma committed:
Bump version

Commit:
fuzz: add fuzzing harness for the native and ssh formats

o builds and fuzzes in a github action flow.

Commit:
tests: also free alloced memory before exiting

This makes LeakSanitizer happy

Commit:
Check that strtok_s succeeded before continuing

Avoids a potential segfault found by fuzzing

Commit:
docs: cosmetics

a-dma committed:
Merge pull request #157 from Yubico/fuzzi

Some minor things found by fuzzing

a-dma committed:
Merge pull request #158 from Yubico/docs

docs: cosmetics

Commit:
Github Actions: build and run tests on linux

Commit:
Merge pull request #159 from Yubico/ga

Github Actions: build and run tests on linux

a-dma committed:
Do not error out when nodetect is set and devices are found

a-dma committed:
Merge pull request #162 from Yubico/nodetect_fix

Do not error out when nodetect is set and devices are found

Closes #161

Commit:
Merge pull request #160 from Yubico/fuzz

Add fuzzing harness for fuzzing native and ssh formats

a-dma committed:
Add a note regarding SELinux on man and README

Relates to #152.

a-dma committed:
Clarify text around SELinux

Relates to #152.

sgn committed:
b64,util: always include limits.h

UCHAR_MAX and INT_MAX are defined within limits.h

Current codebase failed to be built in Linux with musl libc due to
missing this inclusion.

Fix it.

a-dma committed:
Merge PR #165

Commit:
actions: add xsltproc as dependency

Commit:
actions: apt update

andrewkozlik committed:
Support self-attestation in pamu2fcfg.

a-dma committed:
Merge PR #168

a-dma committed:
Bump minimum required version of libfido2

PR #168 added support for self-attestation in pamu2fcfg which was
introduced in libfido2 v1.3.0.

loshz committed:
files -> authMappingFiles

a-dma committed:
Merge pull request #169 from syscll/syscll-patch-1

README: update Authorization Mapping Files link

LDVG committed:
README: update dependency installation instructions

LDVG committed:
git: ignore dirstamp files created by automake

LDVG committed:
Merge pull request #174 from Yubico/readme

misc: update README, gitignore

martelletto committed:
Handle converse() returning NULL

If a PIN is required and converse() returns NULL, abort the
authentication flow instead of reverting to FIDO2 without PIN.
Fixes #175.

martelletto committed:
Verify that the UV bit is set

If PIN or other forms of user verification are required, check that
the corresponding User Verification (UV) bit is set in the signed
payload obtained from the authenticator.

LDVG committed:
fuzz: use COMPILER_CLANG conditional to add subdir

Otherwise, the fuzz subdirectory is not included in the tarball created
by `make dist`. If one then tries to build from the tarball with
CC=clang, then `./configure` fails fails with

    config.status: error: cannot find input file: `fuzz/Makefile.in'

LDVG committed:
fuzz: also include helper scripts in tarball

LDVG committed:
Update NEWS file

LDVG committed:
Bump version

LDVG committed:
man: clarify documentation about authfile path

This resolves #182.

LDVG committed:
ci: xenial is no longer supported

LDVG committed:
ci: add supported ubuntu distros to workflow

Unfortunately, only current LTS releases are provided by GitHub.

LDVG committed:
ci: install libfido2 via ppa for ubuntu-18.04

LDVG committed:
ci: workaround for libfido2 1.7.0 pkg-config

LDVG committed:
ci: bump actions/checkout to v2

LDVG committed:
ci: add format check (previously on travis)

LDVG committed:
ci: add macos workflow

LDVG committed:
ci: retire travis builds

LDVG committed:
make: check whether linker supports -Wl,--wrap

This adds the AX_CHECK_LINK_FLAG macro. Comfortably makes the build work
on MacOS+clang without adding additional configure flags.

LDVG committed:
ci: increase number of 'make' jobs

LDVG committed:
ci: remove checkout of 'HEAD^2', as per build annotation

LDVG committed:
ci: replace old badges in readme with github actions badges

LDVG committed:
ci: run scan-build on builds with clang

LDVG committed:
gitignore: ignore build/ directory

LDVG committed:
util: remove unused return value

To appease scan-build.

LDVG committed:
util: zap unnecessary retval variable

LDVG committed:
make: use correct variables for AM_CFLAGS

This project doesn't use warnings/manywarnings since ed1b7c74.

LDVG committed:
pamu2fcfg: remove unused variables

LDVG committed:
pamu2fcfg: reduce scope of iterator variable

This also fixes warnings about shadowing the variable via another
for-loop in the same function.

LDVG committed:
pamu2fcfg: always explicitly free allocated memory

LDVG committed:
pamu2fcfg: close any opened device

LDVG committed:
pamu2fcfg: free memory allocated by cmdline_parser()

LDVG committed:
pamu2fcfg: introduce print_authfile_line()

LDVG committed:
pamu2fcfg: update copyright

LDVG committed:
pamu2fcfg: introduce verify_cred()

LDVG committed:
pamu2fcfg: introduce make_cred()

perceival committed:
Update README with Fedora dependencies

LDVG committed:
pamu2fcfg: introduce prepare_cred()

LDVG committed:
make: improve checks for compiler warnings under clang

By default, clang only warns for unknown warnings. This makes the
AX_CHECK_COMPILE_FLAG macro always include the flags, regardless whether
they are supported. To remedy, add -Werror=unknown-warning-option.

LDVG committed:
pamu2fcfg: add strlcpy() via compat layer

LDVG committed:
pamu2fcfg: make origin an array, drop generic buf

LDVG committed:
ci: add musl build via alpine container

LDVG committed:
tests: fix memset with sizeof of incorrect type

LDVG committed:
pamu2fcfg: have clang-format ignore imported file

LDVG committed:
man: mirror readme's authfile description with man pages

However, remove the part about the authfile format since it is actually
explained in the linked sections.

LDVG committed:
man: update authfile format

perceival committed:
Update README

Fixed missing Fedora dependencies

LDVG committed:
readme: reword and reorder fedora instructions

For consistency with the other install instruction.

LDVG committed:
Merge pull request #188 from perceival/patch-1

Update README with Fedora dependencies

SoapGentoo committed:
respect `DESTDIR`

* Without `DESTDIR`, `make install` will try to delete files from the live filesystem.
* Also, do not change `libdir`, create a separate prefix for the pam plugin instead.

LDVG committed:
Merge pull request #193 from SoapGentoo/respect-DESTDIR

respect `DESTDIR`

LDVG committed:
make: add configure flag for enabling fuzzing targets

Prefer explicitly enabling/disabling fuzzing over detecting compiler/support.

LDVG committed:
tests: always memset entire dev array

LDVG committed:
util: plug memory leak when using old format pubkey

LDVG committed:
util: remove unused client_key attribute

LDVG committed:
util: introduce set of functions to initialize an assert

LDVG committed:
util: introduce helpers for managing options

Note that this changes how the manual authentication paths parses
authenticator options, now taking into account any options set in the
module configuration. These were not considered before.

LDVG committed:
util: set of functions for managing public keys

LDVG committed:
util: more use of is_resident()

LDVG committed:
util: introduce manual_get_assert()

LDVG committed:
pamu2fcfg: use cose_type(), introduce cose_string()

LDVG committed:
pamu2fcfg, util: add support for EDDSA

LDVG committed:
man: document support for EDDSA

LDVG committed:
util: case-insensitive cose_type()

pamu2fcfg documentation expresses cose type in uppercase, preserve
compatibility using strcasecmp (which is what pamu2cfg did before).

LDVG committed:
util: pull in cose_string() from pamu2fcfg

LDVG committed:
util: introduce load_ssh_key

This parses the SSH key file and returns the base64-encoded key.

LDVG committed:
utils: introduce set of helpers for parse_ssh_format()

Among other things, this ensures that any strings that we log to the
debug file are null-terminated.

LDVG committed:
util: add support for sk-ssh-ed25519

LDVG committed:
util: introduce reset_device()

LDVG committed:
util: decorate _debug with the format attribute

LDVG committed:
fuzz: wrap more symbols, reorganize linker flags

This wraps most of the external symbols used by pam_u2f.la. Note that
the implementation of wrap.c depends on the deterministic prng included
with fuzzing-instrumented builds of libfido2 (-DFUZZ=1).

LDVG committed:
fuzz: add fuzz_auth harness

This harness fuzzes the entirety of pam-u2f via the
pam_sm_authenticate() entry point.

LDVG committed:
pam-u2f: handle multiple occurences of the `debug_file` argument

If pam-u2f is given the following configuration

    debug_file=/some/file debug_file=syslog

it will ultimately try calling

    fclose((FILE *) -1);

and promptly crash. Instead, close any potentially opened file and
reset the configuration whenever pam-u2f parses any occurence of the
`debug_flag` argument.

LDVG committed:
pam-u2f/fuzz: while fuzzing, restrict maximum number of devices

Otherwise, libfuzzer generates OOM errors.

LDVG committed:
util: prevent misaligned loads when parsing ssh key

LDVG committed:
ci/fuzz: fuzzing-compatible builds of libfido2,libcbor

Make it possible for mocked devices' wiredata to propagate through
libfido2 (see nonce, cid, et al).

LDVG committed:
pam-u2f: free response after interactive prompt

LDVG committed:
fuzz: use seed type suitable for the left shifts

LDVG committed:
fuzz: tweak coverage.sh for project-wide coverage instrumentation

LDVG committed:
README: use console-style source highlighting

Missing language specifiers rendered code blocks invisible on
developers.yubico.com. Using `bash` as the langauge specifier leads to
commands prepended with a hash symbol being treated as comments.
Instead, use the `console` specifier.

LDVG committed:
util: handle old format with origin != appid

This fixes an compatibility problem that occurred when pam-u2f
switched to libfido2 (v1.1.0). Users with old-format authfile entries
with differing `origin` and `appid` strings were no longer able to
authenticate as only the `origin` is used. For context, see GitHub
issue #205.

LDVG committed:
ci: update fuzz corpus url and paths

LDVG committed:
ci: replace nproc call on macos builds

LDVG committed:
priv: use openpam_{borrow,restore}_cred when available

Provide these functions as aliases to pam_modutil_{drop,regain}_priv.
While here, move `configure.ac` to use `AC_CHECK_FUNCS()` for PAM
extensions.

This resolves #207.

LDVG committed:
priv: drop bespoke implementation of pam_modutil_{drop,regain}_priv

linux-pam have had these functions implemented since 2010. The
alternatives provided by openpam have been around since 2002.

LDVG committed:
readme: add reference to pam.conf(5)

LDVG committed:
readme: add a warning about erroneous pam configurations

ghost commented:
## DeepCode's analysis on [#5fb6ac](https://www.deepcode.ai/app/gh/cbk914/pam-u2f/65a4ac96bcf2ecd8dcfe9f188a80832d44e5b265/cbk914/pam-u2f/5fb6acece0dfb5382374e90abc87b5e4b8cb732c/pr/_/%2F/code/?utm_source=gh_review&c=0&w=2&i=1&) found:

- :warning: **2** warnings, :information_source: **1** minor issue. :point_down:
- :heavy_check_mark: **2** issues were fixed.
## Top issues

<table>
  <thead>
    <tr>
      <th align="left">Description</th>
      <th align="left">Example fixes</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td width="77%">Usage of insecure temporary file. If an attacker is able to guess the filename, he might be able to create the file with less restrictive access rights and gain access to the contents written by the application. Consider using the safer alternative mkstemp. Occurrences: <ul><li><a href="https://github.com/cbk914/pam-u2f/blob/5fb6acece0dfb5382374e90abc87b5e4b8cb732c/fuzz/fuzz_format_parsers.c#L76">fuzz_format_parsers.c:76</a></li></ul></td>
      <td width="23%"><a href="https://www.deepcode.ai/app/gh/cbk914/pam-u2f/65a4ac96bcf2ecd8dcfe9f188a80832d44e5b265/cbk914/pam-u2f/5fb6acece0dfb5382374e90abc87b5e4b8cb732c/pr/_/%2Ffuzz%2Ffuzz_format_parsers.c/cpp%2Fdc%2FInsecureTmpFile/code/?utm_source=gh_review&">:wrench: Example fixes</a></td>
    </tr>
    <tr>
      <td width="77%">The expression will always evaluate to false because both sides always hold the same value. Occurrences: <ul><li><a href="https://github.com/cbk914/pam-u2f/blob/5fb6acece0dfb5382374e90abc87b5e4b8cb732c/pamu2fcfg/pamu2fcfg.c#L149">pamu2fcfg.c:149</a></li></ul></td>
      <td width="23%"><a href="https://www.deepcode.ai/app/gh/cbk914/pam-u2f/65a4ac96bcf2ecd8dcfe9f188a80832d44e5b265/cbk914/pam-u2f/5fb6acece0dfb5382374e90abc87b5e4b8cb732c/pr/_/%2Fpamu2fcfg%2Fpamu2fcfg.c/cpp%2Fdc%2FCppSameEvalBinaryExpressionfalse/code/?utm_source=gh_review&">:wrench: Example fixes</a></td>
    </tr>
    <tr>
      <td width="77%">Converting integer types to pointer types can lead to portability issues and unspecified behavior if the pointer and integer types become of different sizes. Occurrences: <ul><li><a href="https://github.com/cbk914/pam-u2f/blob/5fb6acece0dfb5382374e90abc87b5e4b8cb732c/pam-u2f.c#L99">pam-u2f.c:99</a></li></ul></td>
      <td width="23%"><a href="https://www.deepcode.ai/app/gh/cbk914/pam-u2f/65a4ac96bcf2ecd8dcfe9f188a80832d44e5b265/cbk914/pam-u2f/5fb6acece0dfb5382374e90abc87b5e4b8cb732c/pr/_/%2Fpam-u2f.c/cpp%2Fdc%2FCastIntegerToAddress/code/?utm_source=gh_review&">:wrench: Example fixes</a></td>
    </tr>
  </tbody>
</table>


#### 👉 View analysis in [**DeepCode’s Dashboard**](https://www.deepcode.ai/app/gh/cbk914/pam-u2f/65a4ac96bcf2ecd8dcfe9f188a80832d44e5b265/cbk914/pam-u2f/5fb6acece0dfb5382374e90abc87b5e4b8cb732c/pr/_/%2F/code/?utm_source=gh_review&c=0&w=2&i=1&) | [_Configure the bot_](https://www.deepcode.ai/app/gh/?ownerconfig=cbk914)
#### 👉 The DeepCode service and API will be deprecated in August, 2021. [Here](https://www.deepcode.ai/blog/migrate-to-snyk-code) is the information how to migrate. Thank you for using DeepCode 🙏 ❤️ !
If you are using our plugins, you might be interested in their successors: [Snyk's JetBrains plugin](https://snyk.co/udpkq) and [Snyk's VS Code plugin](https://snyk.co/udpkr).

LDVG committed:
util: filter authenticators based on selected options

If the user specifies PIN or user verification, only proceed with CTAP2
authenticators that support these modes. Moreover, if user verification
is disabled and unsupported by the authenticator, omit the option
instead of explicitly specifying it (as per spec).

This can be simplified when we can enforce libfido2 1.7.0 or greater
using the fido_dev_{supports,has}_{pin,uv} family of functions.

LDVG committed:
man: improve documentation on origin/appid

Since pam-u2f/pamu2fcfg 1.1.0, these are the RP ID and RP name
respectively.

LDVG committed:
ci: revert workaround for libfido2 1.7.0 pkg-config

libfido2 has been updated to 1.8.0 and fixed its pkg-config file.

This reverts commit f704ec7ff92a84f7adfcf6564e4ff6ec26f3b19e.

LDVG committed:
make: support distcheck without sudo

When distcheck runs, the packages are installed in two temporary staging
directories (once with --prefix, once with DESTDIR) and then removed
again. Since we are using `--with-pam-dir` to determine where to install
pam_u2f.so, we must help distcheck with an extra flag.

LDVG committed:
make: respect DESTDIR on uninstall

LDVG committed:
ci: also run distcheck

LDVG committed:
Update version number to reflect changes

LDVG committed:
Update NEWS file

LDVG committed:
Bump version

LDVG committed:
pam: add helper function for `interactive` prompt

Reduces scope of helper variable required to free the user response.

LDVG committed:
pam: add helper function to resolve `authfile` path

Given a relative or unassigned authfile path, this function allocates
and returns the absolute path to the authfile.

LDVG committed:
util: assume that syslog is defined

Windows is not a target for us, syslog should always be available.

LDVG committed:
util: use NULL to represent syslog

While here, bump copyright.

LDVG committed:
pam: introduce helpers for managing the `debug_file`

LDVG committed:
debug: introduce debug.{c,h}

Move debug-related functions here and introduce debug_dbg() which can be
used to replace if-debug-debug constructs.

LDVG committed:
pam: replace D() calls in parse_cfg()

LDVG committed:
pam: replace DBG() calls with debug_dbg()

LDVG committed:
util: use debug_dbg() in native authfile parser

LDVG committed:
util: use debug_dbg() in SSH authfile parser

LDVG committed:
util: use debug_dbg() in authfile loader

LDVG committed:
util: use debug_dbg() while probing authenticators

LDVG committed:
util: use debug_dbg() while preparing assert

LDVG committed:
util: use debug_dbg() in do_authentication()

LDVG committed:
util: use debug_dbg() in do_manual_authentication()

LDVG committed:
debug: only write to debug file once per call

This prevents double syslog entries for each call of `debug_fprintf()`
by writing into an internal buffer before passing along the result.
This may truncate long debug messages (>2048 octets; the recommended
minimum maximum syslog message length), which should be avoided. Note
that RFC 5424 §8.3 encourages not logging messages longer than the
required minimum maximum size of 480 octets.

LDVG committed:
debug: only print the source file's basename

The file path may contain an absolute or relative path to the source
directory. Strip any components before the last forward slash.

LDVG committed:
util: reword potentially long debug messages

Move user-controlled data towards the end of the debug string so that
relevant information is always visible. Additionally remove a couple of
overly verbose debug messages.

LDVG committed:
debug: indicate truncated log messages

LDVG committed:
m4: update ax_check_compile_flag to latest revision

LDVG committed:
autoconf: replace obsoleted macros

- AC_PROG_CC_STDC was integrated into AC_PROG_CC in autoconf 2.54.
- Replace only remaining occurrence of obsolete AC_HELP_STRING.
- Modern libtool prefers passing options through LT_INIT.

LDVG committed:
COPYING: bump copyright

martelletto committed:
git -> https

the unauthenticated git protocol on port 9418 (git://) is no
longer supported by github.

LDVG committed:
git: simplify .gitignore

Remove ignores of specific files that are also ignored by patterns.

LDVG committed:
man: collect man page generation rules

`DISTCLEANFILES` is replaced with `MAINTAINERCLEANFILES`, the former is
primarily intended for files generated by `configure`. We are
distributing the generated man pages as if they were source files.

LDVG committed:
pamu2fcfg: prevent parallel generation of cmdline.[ch]

Running make in parallel would spawn two instances of gengetopt. To
prevent this, use a GNU make pattern rule which is able to express rules
with multiple output files. Other make implementations should still be
able to build from the tarballs. While here, remove an unnecessary
include path.

LDVG committed:
tests: use convenience library for unit tests

Tests that poke at internal functions can link to this (static) library
instead. This is required if we're to restrict the symbols that are
exported in our shared module.

LDVG committed:
pam: only export the public service functions

For fuzzing, a couple of additional symbols for mocking are exported.

LDVG committed:
tests: basic sanity checks of the shared module

This checks that the expected public service functions can be loaded
using `dlsym`.

LDVG committed:
tests: fixup CFLAGS and related warnings

LDVG committed:
pam: remove unused variable

LDVG committed:
pamu2fcfg: use bespoke `getopt_long` handler

This drops the (maintainer) dependency on `gengetopt`. Moreover, a
combination of the `--username` and `--nouser` options is now allowed.
The former sets the username in the credential, while the latter only
toggles whether the username is printed to stdout with the rest of the
credential.

LDVG committed:
doc: remove occurrences of `gengetopt`

LDVG committed:
ci: remove occurrences of `gengetopt`

alexgeana committed:
fuzz: checkout specific libfido2 commit for fuzzing builds

LDVG committed:
tests: simplify regenerate_credentials.py

Reduce repetition via a helper function to generate and dump credentials
to the specified filename.

LDVG committed:
tests: modernize regenerate_credentials.py for PY3

LDVG committed:
tests: flatten regenerate_credentials.py loops

LDVG committed:
pam: set asprintf string to NULL on failure

LDVG committed:
pam: remove unnecessary branch

Caller guarantees a relative or missing path.

LDVG committed:
pam: reduce `openasuser` repetition

LDVG committed:
pam: reduce `asprintf` repetition

While here, ensure that the resolved path is absolute.

LDVG committed:
fuzz: sort includes

LDVG committed:
fuzz: wrap asprintf

LDVG committed:
tests: only regenerate missing templates

While here, remove redundant argument.

LDVG committed:
tests: clearly label autogenerated test cases

Additionally, skip running the formatter on the autogenerated code to
make it easier to diff in the future. The two tests that are not
automatically generated (SSH, old format) are kept separate.

While here, remove unnecessary includes.

alexgeana committed:
fuzz: check prng status before use

LDVG committed:
pam: skip redundant configuration option comparisons

No functional change; each option has a unique name.

LDVG committed:
pam: use strcmp for sshformat flag matching

This ensures an exact match of the configuration option. Prefix matching
is only useful for matching configuration options with a value.

LDVG committed:
pam: remove redundant appid comparison

No functional change; it does not matter if we pass `origin` or `appid`
to `fido_assert_set_rp()` if they compare equal.

LDVG committed:
configure: remove redundant variables

AC_CHECK_FUNCS does not modify CFLAGS nor LIBS, there's no need to store
and reset their value. Remove redundant comment.

LDVG committed:
configure: remove unused header check

LDVG committed:
fuzz: sort headers

LDVG committed:
fuzz: export get_devices_from_authfile()

Have `fuzz_format_parsers` use this function instead of the two private
authfile parsers. This is made possible via the mocking utilities
introduced along with `fuzz_auth` and enables us to not have to include
the relevant sources directly in the fuzzer harness source.

alexgeana committed:
tests: remove deprecated scan.yml workflow

LDVG committed:
pam: introduce parse_native_credential()

extracts (native) credential components from a comma-delimited list

LDVG committed:
ci: fix permissions on alpine workdir

LDVG committed:
pam: ensure reaching native authfile's EOF

since we look for the user's last line in the authfile, ensure that all
of it was parsed when fgets() or friends returns without data

LDVG committed:
pam: consolidate authfile parsers' return values

other functions in pam-u2f return 1 for success, 0 for error

LDVG committed:
pam: simplify ssh authfile parser cleanup

LDVG committed:
pam: simplify authfile parser cleanup

LDVG committed:
pam: do not truncate credentials

ensure an entire line is always read as to not truncate parts of the
credential

LDVG committed:
make: silence deprecation warnings for OSSL 3.*

When we bump up the minimum libfido2 version to 1.9.0 we can switch to
using the EVP interface rather than the deprecated EC_KEY interface.
Until then, silence the related deprecation warnings.

LDVG committed:
pam: adjust buffer size correctly

only decrement size when we copy a character

LDVG committed:
pam: use fixed max size for SSH credential

We're always parsing at most one credential. Use a reasonably sized
buffer to parse into.

LDVG committed:
pam: mandate a SSH trailer for success

LDVG committed:
pam: remove unused buffer

MartinBrugnara committed:
FAQ: YubiKey 4 does not support asking for pin

Explicitly state YubiKey 4 series does not support asking for pin.

__Problem__
* FIDO(1) devices are only partially supported. They can be used as a second authentication factor to the password or as a single factor by themselves, but they cannot be used in conjunction with a `pin`.

* The failure mode consists in silently ignoring the device.
It does not provide any feedback as to why the authentication procedure is failing and it's thus very complicated to debug (see for example #144).

* YubiKey product pages, understandably, do not highlight the differences between FIDO and FIDO2, users may thus not realize at a first glance what their devices do actually support.

__This PR__
Add a paragraph to help users like myself understand why their YubiKeys work only half of the time.

__Next steps__
Modify `pamu2fcfg` to emit a warning when registering devices with unsupported configurations.

LDVG committed:
man: reword, rewrap previous

LDVG committed:
man: make note of features that require FIDO2

LDVG committed:
man: clarify omission of options

LDVG committed:
fuzz: wrap getline

LDVG committed:
tests: test parsing a limited number of credentials

LDVG committed:
man: sync manpages with pamu2fcfg help text

LDVG committed:
Update NEWS file

LDVG committed:
Bump version

LDVG committed:
ci: update codeql to v2

LDVG committed:
ci: explicitly add bash to alpine container

LDVG committed:
ci: bump alpine host to ubuntu-22.04

LDVG committed:
ci: alpine now packages libfido2-dev

LDVG committed:
ci: collapse linux build matrix

LDVG committed:
ci: add ubuntu 22.04 builds

LDVG committed:
ci: bump fuzzer builds to ubuntu 22.04/clang-14

LDVG committed:
ci: add macos-11, macos-12

LDVG committed:
fuzz: bump libfido2 to 1.11.0

LDVG committed:
fuzz: bump libcbor to 0.9.0

LDVG committed:
tests: prefer getpwuid over getenv

`$USER` may not be set, look up the username of the effective user ID
from the password database instead. In the long term, consider splitting
authfile handling and parsing to simplify testing.

This fixes #270.

LDVG committed:
ci: drop USER environment variable
